[libc] cmake configruation for newhdrgen

aaryanshukla opened the pull request:
- this is a draft to get more input/reviews
- tests have been done on dirent and ctype to see matching results as
  old headergen
- added entrypoints filtering within the script
- will switch function name to add_newgen_header

aaryanshukla committed:
[libc] cmake configruation for newhdrgen

- this is a draft to get more input/reviews
- tests have been done on dirent and ctype to see matching results as
  old headergen
- added entrypoints filtering within the script

lntue commented on libc/cmake/modules/LLVMLibCHeaderRules.cmake:
```diff
@@ -66,6 +67,90 @@ function(add_header target_name)
+  set(entry_points "${TARGET_ENTRYPOINT_NAME_LIST}")
+  list(TRANSFORM entry_points PREPEND "--e=")
+
+  add_custom_command(
```
Don't use `add_custom_command` and depending on its output file.  Just do `add_custom_target` with a `COMMAND` field for easier to manage dependency.

aaryanshukla replied:
@lntue fixed it!

aaryanshukla committed:
swapped custom_command with custom_target

michaelrj-google commented on libc/cmake/modules/LLVMLibCHeaderRules.cmake:
```diff
@@ -66,6 +67,90 @@ function(add_header target_name)
   )
 endfunction(add_header)
 
+function(add_yaml_header target_name)
```
nit: This shouldn't be named `add_yaml_header` since the header itself doesn't have any yaml. Maybe `add_gen_header_2`?

aaryanshukla replied:
@michaelrj-google add_newgen_header?

michaelrj-google commented on libc/cmake/modules/LLVMLibCHeaderRules.cmake:
```diff
@@ -66,6 +67,90 @@ function(add_header target_name)
+    DEPENDS ${yaml_file} ${def_file} ${fq_data_files}
+    COMMENT "Generating header ${ADD_YAML_HDR_GEN_HDR} from ${yaml_file} and ${def_file}"
+  )
+
```
it looks like in the old version the GPU build had an extra condition here that created a second set of headers that only contained the function prototypes. @jhuber6 do you still need that?

michaelrj-google reviewed (commented):
Overall looks good, but wait for @jhuber6 to weigh in on the GPU side

aaryanshukla committed:
switched back to custom_command

-  able to generate all headers with new headergen  (waiting on landing
   for assert.yaml)
- fixed bugs within yaml files
